### ISSUE TITLE
Document batch_norm2d pathological test case in testing guide

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#3282]
The `grad_output = ones_like(output)` case for batch norm produces analytically-zero gradients due to the normalization property `sum(x_norm) = 0`. This caused a false ~1000x mismatch that was mistaken for a bug. This edge case should be documented in `docs/dev/testing-strategy.md` or the backward-pass-catalog as a known gotcha for testing batch norm backward passes.

_Follow-up from #2724_

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have enough context. Here's the implementation plan:

---

## Implementation Plan: Issue #3282 — Document batch_norm2d Pathological Test Case

### 1. Objective

Document the `grad_output = ones_like(output)` pathological test case for `batch_norm2d` backward passes. When `beta=0`, this produces analytically-zero gradients due to the normalization property `sum(x_norm per channel) = 0`, causing a false ~1000x mismatch that looks like a bug but is mathematically correct. This gotcha must be captured in `docs/dev/testing-strategy.md` so future implementers don't repeat the investigation.

---

### 2. Approach

- **Single-file documentation change**: Add a new "Known Gotchas" section to `docs/dev/testing-strategy.md`, placed after the existing "Gradient Checking" section. This is the right home because the issue is about test design, not the backward-pass formula itself.
- **Also add a brief cross-reference note** in `docs/dev/backward-pass-catalog.md` under the batch norm entry (once confirmed it exists or is added).
- No code changes — this is purely documentation.

---

### 3. Files to Create

None.

---

### 4. Files to Modify

#### `docs/dev/testing-strategy.md`

Add a new **"Known Test Gotchas"** section after the existing "Gradient Checking" section (currently ending around line 214). The section should contain:

- **Subsection**: `BatchNorm2d: ones_like(grad_output) with beta=0 produces zero gradients`
  - The identity: `sum over (b,h,w) of output[b,c,h,w] = gamma * sum(x_norm) + N*beta = N*beta`; with `beta=0` this is always 0 regardless of input x.
  - The Kratzert three-term formula derivation showing why `grad_input = 0` when `grad_output = ones` and `beta = 0`:
    - `k = N` (sum of ones)
    - `dotp = sum(x_norm) = 0` (zero-mean property of batch norm in training mode)
    - Result: `grad_input[i] = (1 - N/N - x_norm[i] * 0/N) * gamma/std = 0`
  - Why the numerical gradient also appears to be ~0 (the loss function `sum(output)` is degenerate when `beta=0`)
  - The false ~1000x mismatch artifact: noise in finite-difference estimation vs exact zero
  - **Safe alternatives**:
    - Use `beta != 0` so `sum(output) != 0`
    - Use `grad_output = random` instead of `ones_like`
    - Use a non-trivial loss (e.g., `sum(output^2)`)
  - Code example showing the safe pattern

#### `docs/dev/backward-pass-catalog.md`

Add a **"Known Test Pathologies"** subsection to the BatchNorm backward entry (or append one if the entry is missing). Cross-reference the testing-strategy gotcha section.

---

### 5. Implementation Order

1. Read `docs/dev/backward-pass-catalog.md` fully to find the BatchNorm section (or confirm it's missing).
2. Edit `docs/dev/testing-strategy.md`:
   - Insert new "Known Test Gotchas" section after the Gradient Checking section (after line ~214).
   - Include the mathematical proof, false-failure explanation, and safe test patterns.
3. Edit `docs/dev/backward-pass-catalog.md`:
   - Add or update the BatchNorm2d backward entry with a cross-reference to the gotcha in testing-strategy.md.
4. Run markdown linting: `just pre-commit-all` (or `pixi run npx markdownlint-cli2 docs/dev/testing-strategy.md docs/dev/backward-pass-catalog.md`).
5. Fix any linting issues (blank lines around code blocks/headings, line length ≤120 chars).
6. Commit on the current branch `3013-auto-impl` and create/update a PR referencing issue #3282.

---

### 6. Verification

- `pixi run npx markdownlint-cli2 docs/dev/testing-strategy.md` passes with no errors.
- `pixi run npx markdownlint-cli2 docs/dev/backward-pass-catalog.md` passes with no errors.
- `just pre-commit-all` passes (trailing whitespace, end-of-file, YAML, line-endings).
- A reviewer reading the gotcha section can understand: (a) why the zero-gradient result is correct, (b) how to avoid the pathology in new tests.

---

### 7. Content Details for the Gotcha Section

The section to insert into `docs/dev/testing-strategy.md` after the Gradient Checking section:

```markdown
## Known Test Gotchas

This section documents edge cases that have caused false failures in the past.
Always check here before assuming a gradient mismatch is a real bug.

### BatchNorm2d: `ones_like(grad_output)` with `beta=0` Produces Zero Gradients

**Symptom**: Analytical gradient ≈ 0, numerical gradient ≈ 0.009 — apparent ~1000x mismatch.

**Root cause**: This is **not a bug**. The zero analytical gradient is mathematically correct,
and the numerical gradient is also essentially zero — the apparent mismatch is finite-difference
noise against an exactly-zero baseline.

#### Mathematical Proof

Batch normalization normalizes each channel over the batch and spatial dimensions:

```text
x_norm[b,c,h,w] = (x[b,c,h,w] - mean_c) / std_c
output[b,c,h,w] = gamma_c * x_norm[b,c,h,w] + beta_c
```

The key identity is the **zero-mean property**: for any input x,
the normalized values per channel always sum to zero in training mode:

```text
sum over (b,h,w) of x_norm[b,c,h,w] = 0  (by definition of mean-subtraction)
```

Therefore:

```text
sum over (b,h,w) of output[b,c,h,w]
    = gamma_c * sum(x_norm) + N * beta_c
    = gamma_c * 0 + N * beta_c
    = N * beta_c
```

With `beta = 0`: **`sum(output) = 0` for any input x**.

#### Why the Gradient is Zero

Using the Kratzert three-term batch norm backward formula:

```text
grad_input[i] = (grad_output[i] - k/N - x_norm[i] * dotp/N) * gamma/std
```

where `k = sum(grad_output)` and `dotp = sum(grad_output * x_norm)`.

When `grad_output = ones_like(output)`:

- `k = N` (sum of N ones)
- `dotp = sum(1 * x_norm) = sum(x_norm) = 0`  ← zero-mean property
- Result: `grad_input[i] = (1 - N/N - x_norm[i] * 0/N) * gamma/std = 0`

This is **analytically exactly zero**, not approximately zero.

#### Why Numerical Gradient Also Fails

The test loss `L = sum(output)` is itself degenerate when `beta=0`:
`L = N * beta = 0` identically, so numerical finite differences also return ~0
(up to floating-point noise). The "0.009" discrepancy is pure numerical noise,
not a real gradient value.

#### Safe Alternatives

**Option 1**: Set `beta != 0` so `sum(output) != 0`

```mojo
var beta = create_constant_tensor([C], DType.float32, 0.1)  # Non-zero beta
```

**Option 2**: Use random `grad_output` instead of `ones_like`

```mojo
var grad_output = create_seeded_random_tensor(output.shape(), DType.float32, seed=42)
```

**Option 3**: Use a non-trivial loss (sum of squared outputs)

```mojo
# Loss = sum(output^2), grad_output = 2 * output
var grad_output = mul_scalar(output, 2.0)
```

#### Discovery Context

Found during PR #2724 / issue #3282. Implemented fix: replaced `TODO: still investigating`
skip with the direct function call once the math was verified. See `batch-norm-backward-gradient-analysis`
in the team knowledge base for the full investigation.
```

---

### 8. Skills Used

| Skill | How Used |
|-------|----------|
| `batch-norm-backward-gradient-analysis` *(team knowledge base)* | Primary source of the mathematical proof, false-failure analysis, and safe alternatives documented in this plan |
| `fix-ci-mojo-test-failure` *(team knowledge base)* | Referenced for the false-failure investigation pattern |
| `Read` | Read `docs/dev/testing-strategy.md` and `docs/dev/backward-pass-catalog.md` to understand existing structure and insertion points |
| `Glob` | Located the backward-pass-catalog file path |
| `Grep` | Searched for existing BatchNorm coverage in both docs files |

---
*Generated by Claude Code Planner*
